src/krb5/_localauth.pyx

```diff
@@ -0,0 +1,47 @@
+# Copyright: (c) 2022 Jordan Borean (@jborean93) <jborean93@gmail.com>
+# MIT License (see LICENSE or https://opensource.org/licenses/MIT)
+
+from libc.stdlib cimport free, malloc
+
+from krb5._exceptions import Krb5Error
+
+from krb5._context cimport Context
+from krb5._krb5_types cimport *
+from krb5._principal cimport Principal
+
+
+cdef extern from "python_krb5.h":
+
+    krb5_error_code krb5_aname_to_localname(
+        krb5_context context,
+        krb5_const_principal aname,
+        int lnsize,
+        char *lname_out
+    ) nogil
+
+    krb5_error_code KRB5_LNAME_NOTRANS
+
+def aname_to_localname(
+    context: Context,
+    principal: Principal
+) -> str:
+
+    cdef krb5_error_code err = 0
+    limit = 256
+
+    cdef char *localname = <char *>malloc(limit + 1)
+    if not localname:
+        raise MemoryError()
+
+    err = krb5_aname_to_localname(context.raw, principal.raw, limit, localname)
+    # The definition of KRB5_LNAME_NOTRANS in the Heimdal header file
+    # included in the pykrb5 source appears to be out of date; this is
+    # the value that actually gets returned in the test -- so I'm just
+    # adding it here for now.
+    if err in [KRB5_LNAME_NOTRANS, -1765328227]:
+        return None
+    elif err != 0:
+        free(localname)
```

Comment:
You free it here but not in the good case, we should always free any memory we allocate or any memory allocated by the krb5 library.

See [kt_default_name](https://github.com/jborean93/pykrb5/blob/f36e82ffa4ddd9f540a0bdf41c13a2d404df284d/src/krb5/_kt.pyx#L377-L408) which has a similar API to this call. It starts with a fixed buffer and continues to increase it until there is enough space. It also ensures the buffer is freed once it has been allocated using a try/finally and uses `strlen` to get the size.

I would see the current code looking like

```python
cdef krb5_error_code err = KRB5_CONFIG_NOTENUFSPACE
buffer_size = 1024
buffer_length = buffer_size
cdef char *buffer = <char *>malloc(buffer_length)
if not buffer:
    raise MemoryError()

try:
    while err == KRB5_CONFIG_NOTENUFSPACE:
        err = krb5_aname_to_localname(context.raw, principal.raw, buffer_length, buffer)

        if err == KRB5_CONFIG_NOTENUFSPACE:
            buffer_length += buffer_size

            # Use a temp var to ensure buffer is always something valid to be freed
            new_buffer = <char *>realloc(buffer, buffer_length)
            if not new_buffer:
                raise MemoryError()
            buffer = new_buffer

        elif err in [KRB5_LNAME_NOTRANS, KRB5_NO_LOCALNAME]:
            return None
            
        elif err:
            raise Krb5Error(context, err)

        else:
            name_len = strlen(buffer)
            return <bytes>buffer[:name_len]

finally:
    free(buffer)
```

I saw in the MIT krb5 and Heimdal tests that they use a buffer size of 1024 and not 256 so I've used that as a base.

Reply:
Thanks for the detailed review! Skimming it everything you say makes sense. I'll try to make the changes in the next few days; I won't get to it right away, I'm afraid.

Reply:
One thing:

> You free it here but not in the good case, we should always free any memory we allocate or any memory allocated by the krb5 library.

I must be missing something: we can't free it in the good case, because we're returning it to the caller. I was assuming that once it's returned it's reified as a Python `str` object and will be managed (garbage-collected) as usual by Python. Is that not the case?

Ah, maybe it's this: an expression like `localname.decode("utf-8")` in my original code, your new `<bytes>buffer[:name_len]`, actually creates a Python object from the C string and returns that -- so now it's OK to free the C buffer before returning. Is that it? Pardon my unfamiliarity with Cython.

Reply:
> I must be missing something: we can't free it in the good case, because we're returning it to the caller. I was assuming that once it's returned it's reified as a Python str object and will be managed (garbage-collected) as usual by Python. Is that not the case?

In that case the use of `localname.decode("utf-8")` was creating a new object anyway. You still need to cleanup the buffer that was malloc'd to avoid memory leaks as there is nothing tracking the lifetime of it. So in both your original code and the one I'm proposing this buffer needs to be freed regardless as the object returned by the function is a brand new copy based on that original buffer.

> I actually thought about the keep-mallocing-until-it's-big-enough strategy, but I didn't because a) I'm suspicious of things that can cause a program to just keep allocating memory endlessly until it crashes

That's a good point, we should probably have a check to see if the `buffer_length > ...`, maybe 8192? You can probably keep the original code if you want but I recommend increasing the buffer size to 1024 rather than 256 as that's what both Heimdal and MIT krb5 uses for it's tests.